timemachine/fe/single_topology.py

```diff
@@ -0,0 +1,784 @@
+    # disabled them, we would *still* get the correct stereochemistry.
+    # consider a benzene devoid of any torsions (proper or improper), or non-bonded
+    # terms, and only angles and bonds are present. The hydrogens would still be correctly
+    # placed since the angles intrinsically restrain them.
+    # (ytz): note this is slightly buggy, multiple period torsions that span the same idxs may contain
+    # a term to planarize
+
+    canonical_ring_bonds = set()
+    for ij in ring_bonds:
+        canonical_ring_bonds.add(dummy.canonicalize_bond(ij))
+
+    planar_bonds_kv = identify_bonds_spanned_by_planar_torsions(proper_idxs, proper_params)
+    canonical_stereo_bonds = set()
+    for k in planar_bonds_kv.keys():
+        if k not in canonical_ring_bonds:
+            src, dst = k
+            src = mol.GetAtomWithIdx(src)
+            dst = mol.GetAtomWithIdx(dst)
+            if src.GetDegree() == 4 or dst.GetDegree() == 4:
+                # if either atom is tetrahedral then it has no pi bonds to conjugate over
+                continue
+            else:
+                canonical_stereo_bonds.add(k)
+
+    return canonical_stereo_bonds
+
+
+def find_stereo_atoms(mol):
+    mol_geom = geometry.classify_geometry(mol)
+    stereo_atoms = set()
+    for a in mol.GetAtoms():
+        a_idx = a.GetIdx()
+        nbs = a.GetNeighbors()
+        if len(nbs) == 4:
+            stereo_atoms.add(a_idx)
+        elif len(nbs) == 3 and mol_geom[a_idx] == LocalGeometry.G3_PYRAMIDAL:
+            # if in ring, or is sulfur or phosphorus, this may not be guaranteed by
+            if a.IsInRing() or a.GetAtomicNum() == 16 or a.GetAtomicNum() == 15:
+                stereo_atoms.add(a_idx)
+        elif len(nbs) > 4:
+            assert 0
+    return stereo_atoms
+
+
+def check_bond_stability(j, k, bond_idxs, bond_params):
+    # tbd this should really be based on the pdf, we should expect to see
+    # near zero probability at bond lengths < 0.5 Angstroms
+    canonical_bonds = dict()
+    for idxs, params in zip(bond_idxs, bond_params):
+        canonical_bonds[dummy.canonicalize_bond(idxs)] = params
+
+    jk = dummy.canonicalize_bond((j, k))
+
+    if jk not in canonical_bonds:
+        return False
+    elif canonical_bonds[jk][0] < 50.0:
+        return False
+    elif canonical_bonds[jk][1] < 0.05:
+        return False
+
+    return True
+
+
+def check_angle_stability(j, k, l, angle_idxs, angle_params):
+
+    canonical_angles = dict()
+    for idxs, params in zip(angle_idxs, angle_params):
+        canonical_angles[dummy.canonicalize_bond(idxs)] = params
+
+    jkl = dummy.canonicalize_bond((j, k, l))
+
+    if jkl not in canonical_angles:
+        return False
+    elif canonical_angles[jkl][0] < 50.0:
+        return False
+    elif abs(canonical_angles[jkl][1] - 0) < 0.05:
+        return False
+    elif abs(canonical_angles[jkl][1] - 3.1415) < 0.05:
+        return False
```

Comment:
I had no idea you could mod over floats in python - let me read up on the behavior more carefully before acceping this commit.

Reply:
Hm - I'm not a huge fan of modding over floats, simple things like `3.5 % 0.2 == 0.2` would return `False` due to round-off. I guess in this case since its a `<` sign it's safe but I'm not sure I'm comfortable in general with float mods.

Reply:
Actually I'm going to just modify the `<0.05` case to throw an assertion. This should never happen.

Reply:
>I'm not a huge fan of modding over floats, simple things like 3.5 % 0.2 == 0.2 would return False due to round-off. I guess in this case since its a < sign it's safe but I'm not sure I'm comfortable in general with float mods.

Huh! I didn't realize mod was an unsafe operation to use here.

If concerned about rounding errors in this comparison, may be preferable to use `np.pi` rather than the literal `3.1415`.

```suggestion
    eq_theta = canonical_angles[jkl][1]
    elif abs(eq_theta - 0) < 0.05:
        return False
    elif abs(eq_theta - np.pi) < 0.05:
        return False
```

>Actually I'm going to just modify the abs(canonical_angles[jkl][1] - 0) case to throw an assertion. This should never happen.

Good call.

(May also want a check that `eq_theta` is in interval [0, pi] -- should never happen, but just to rule out silently returning `True` for `3 * pi` but `False` for `1 * pi`...)

Reply:
this seems reasonable - will add.